### ISSUE TITLE
Support Scala Native

### ISSUE THREAD

**Issue by jdegoes** [zio/zio#639]
Scala Native does not support multi-threading, but that shouldn't be an issue, generally, because ZIO shares CPU between fibers. Just need to make sure no blocking code is used in core (essential) methods and that we don't use data structures / types not supported by Scala Native.

**Comment by rwhaling** [comment]
I just wanted to add some notes here, since folks have asked about this a few times on twitter.  For context, I'm working on an ExecutionContext for scala-native at https://github.com/scala-native/scala-native-loop

I don't quite grok the ZIO internals yet, but it looks like Executor.fromExecutionContext is enough to get a bare minimum Runtime bootstrapped, and LibUV provides enough additional capabilities to align with the scala-js `Runtime[Clock with Console with System with Random]`.  

We're planning to start publishing artifacts for native-loop as soon as SN 0.4.0 drops, so we have a few weeks/months before we can start work on ZIO properly, but we might be able to get a head start on planning at least.

The big question mark is whether there are any JVM data structures/types used that would cause problems.  If there are, it is about 85% likely they can provided/stubbed by copy-and-paste from ScalaJS.  Once we get to that point, I'm happy to help triage/facilitate any missing classes on the SN side.

**Comment by jdegoes** [comment]
@rwhaling Thank you for the head start! I think we will start on this in earnest after we get 1.0 out the door... for sure at the ZIO hackathon in September (if not sooner).

Really excited about and thankful for all the work you do with Scala Native (🙏), and looking forward to helping with the initiative by getting ZIO running on Native.

**Comment by rwhaling** [comment]
So, I spent a couple of hours today hacking on this - I was able to get Zio to compile on SN by removing the specs2 dependencies, but once I got an artifact built locally, I did start to see a linking error on the SN side - looks like a missing toArray method on java.util.Set; I've documented the issue on the SN repo here: 
https://github.com/scala-native/scala-native/issues/1661
This doesn't appear to be a Zio-specific issue, and all of SN's collections code is copied whole from scala.js, so I don't expect this to be a large blocker.

**Comment by softinio** [comment]
Awesome!

Related to the specs2 issue: #1208

**Comment by rwhaling** [comment]
I haven't been able to get Zio itself to build on SN 0.4 yet, but it does look like the underlying linker error is already fixed on SN master, so I've closed the SN issues.  Progress!

**Comment by rwhaling** [comment]
It sounds like y'all are going to be doing awesome stuff with zio-test, but we do have specs2 support for Scala Native now: https://twitter.com/specs2org/status/1158062621796401152

I did also check in with Denys about the SN linker issue - that is fixed on master now, so we should be good to go next time there is a SN release; if I have time in the next few days I will try to build some artifacts locally and test.

**Comment by sideeffffect** [comment]
now we need to wire it up with in the CI, so that it gets tested and released

**Comment by adamgfraser** [comment]
Yes, spent a bit of time working on that. How do I add a cross project dependency for Native? `%%%`? Trying to add the SBT test runner dependency if possible.

**Comment by adamgfraser** [comment]
Okay, I found the Scala Native Test Interface and that seems to be working. Now I am blocked on the `portable-scala-reflect` dependency. It looks like there is an [open ticket](https://github.com/scala-native/scala-native/issues/1279) to support Scala Native but that feature is not available yet. It also looks like adding reflection support in Scala Native is still [forthcoming](https://github.com/scala-native/scala-native/issues/159).

We use reflective instantiation [here](https://github.com/zio/zio/blob/8359a6ab26f0e4ee0854a322d05945227d3b1ffb/test-sbt/shared/src/main/scala/zio/test/sbt/BaseTestTask.scala#L16) to run tests for all specs that extend `AbstractRunnableSpec`.

@rwhaling Do you have any thoughts on the best way to provide the same functionality on Scala Native? Thanks again for all your work on this, by the way!

**Comment by mkotsbak** [comment]
Looks like native libs still are not published to Maven repos.

**Comment by sideeffffect** [comment]
But they are:
 * ZIO https://oss.sonatype.org/content/repositories/snapshots/dev/zio/zio_native0.4.0-M2_2.11/
 * ZIO Streams https://oss.sonatype.org/content/repositories/snapshots/dev/zio/zio-streams_native0.4.0-M2_2.11/
 * ZIO Test https://oss.sonatype.org/content/repositories/snapshots/dev/zio/zio-test_native0.4.0-M2_2.11/

Publishing them was introduced in these PRs: https://github.com/zio/zio/pull/4252, https://github.com/zio/zio/pull/4260.

**Comment by mkotsbak** [comment]
Hmm, maybe I just need to add Sonatype repo as extra then. Didn't find in Maven central repo.

**Comment by sideeffffect** [comment]
Oh, sorry, I haven't explained it very well. The current setup _will_ publish the artifacts to Maven Central. But that will happen only after a new release will be made. So far, only artifacts built from `master` are published, and that means that they will end up in Sonatype _snapshots_, instead of Sonatype _releases_ (i.e. Maven Central).

**Comment by sideeffffect** [comment]
Oh, sorry, I haven't explained it very well. The current setup _will_ publish the artifacts to Maven Central. But that will happen only after a new release will be made. So far, only artifacts built from `master` are published, and that means that they will end up in Sonatype _snapshots_, instead of Sonatype _releases_ (i.e. Maven Central).

**Comment by sideeffffect** [comment]
Oh, actually `1.0.2` is on Maven Central now, even for Native
 * https://repo1.maven.org/maven2/dev/zio/zio_native0.4.0-M2_2.11/
 * https://repo1.maven.org/maven2/dev/zio/zio-streams_native0.4.0-M2_2.11/
 * https://repo1.maven.org/maven2/dev/zio/zio-test_native0.4.0-M2_2.11/

Of course, the artifacts weren't tested and ZIO Test most likely doesn't work, but it's at least available for people to experiment with!

**Comment by mkotsbak** [comment]
Nice, even 1.0.3, but I get an error with Mill:

```
$ mill mill.scalalib.GenIdea/idea
[1/1] mill.scalalib.GenIdea.idea 
Analyzing modules ...
Downloading [2/2] artifacts (~0/0 bytes)
1 targets failed
mill.scalalib.GenIdea.idea java.lang.Exception: Duplicated item inserted into OrderedSet: Failure(
Resolution failed for 1 modules:
--------------------------------------------
  dev.zio:zio-stacktracer_native0.4.0-M2_2.11:1.0.3 
	not found: /home/marius/.ivy2/local/dev.zio/zio-stacktracer_native0.4.0-M2_2.11/1.0.3/ivys/ivy.xml
	not found: https://repo1.maven.org/maven2/dev/zio/zio-stacktracer_native0.4.0-M2_2.11/1.0.3/zio-stacktracer_native0.4.0-M2_2.11-1.0.3.pom
,None)
    mill.api.AggWrapper$Agg$Mutable.append(AggWrapper.scala:56)
```

**Comment by sideeffffect** [comment]
@mkotsbak That is an unfortunate oversight, help is on the way https://github.com/zio/zio/pull/4307

**Comment by mkotsbak** [comment]
Works with snapshots now, with version dev.zio::zio::1.0.3+3-64bea7fb-SNAPSHOT :)

**Comment by sideeffffect** [comment]
ZIO is now compilable with Scala Native 0.4.0, so now the actual work on supporting it can begin.
/cc @lolgab @LeeTibbert

**Comment by LeeTibbert** [comment]
@sideeffffect /cc @lolgab  Thank you for these changes, they will make the next steps of adding
Zio support easier.  I am actively working on providing some prerequisites for Zio and then will
co-ordinate with @lolgab about the main effort.    So there is progress here,  it just has
not come to fruition here.

The two major areas for advancement are 

* porting the SN 0.4.0 Junit testing environment. I've done this
  once, so it should just be time consuming. Should have minimal,
  if any, effect on ZioTest.

* Proving first to myself and then to others that I/we have gotten Scala Native
  to play nicely with Zio fibers. This is the harder of the two and the required
  study is at the Scala Native end.

**Comment by adamgfraser** [comment]
Closing as we support this now.

**Comment by sideeffffect** [comment]
> we support this now

Yes and no. If I recall correctly, we have only preliminary support for Scala Native. Only zio-core and zio-streams work, but zio-test and anything that depends on it (like testing everything, including zio-core itself) doesn't work. We're missing some java.time classes, we're basically blocked on https://github.com/cquiroz/scala-java-time/issues/45

I think it would be better to have this ticket closed only after we can run tests on Scala Native.

/cc @lolgab @ekrich @LeeTibbert

**Comment by adamgfraser** [comment]
That's fine. Happy to reopen.

**Comment by virusdave** [comment]
Any updates / progress on this, or plans to do so, given that zio 2.0 has now been released?

**Comment by sideeffffect** [comment]
WIP. We can now [_build_ tests on ZIO 1](https://github.com/zio/zio/pull/6283). We will be [building them on ZIO 2 in short order](https://github.com/zio/zio/pull/7021).
The remaining challenge is then _running_ them. This has proven a bit more problematic than I've originally envisioned, hence the separation into two steps -- building and only after that running.

**Comment by sideeffffect** [comment]
Quick update on the state of things ZIO Native.

Currently we build and publish most of the important ZIO packages for Native for both ZIO 1 and 2. We also compile tests on both 1 and 2.
@lolgab has already had some fun using ZIO 1 Native with snunit (an HTTP server library):
 * https://github.com/lolgab/snunit-zio-seed.g8
 * https://github.com/lolgab/snunit/tree/0.0.11/snunit-zio/src/snunit/snunitzio

While we're already _compiling_ the tests of ZIO projects, we're currently still not _running_ the tests on ZIO Native. This is the most urgent thing on the roadmap, so if you'd like to help, this would be the best place. Relevant PR:
https://github.com/zio/zio/pull/7031

After that, the next big challenge would be to create a library to abstract I/O on different platforms (JVM, JS, Native). This would allow other libraries in the ZIO space to depend on it while being easily portable. In the TypeLevel/Cats Effect space, [fs2-io](https://fs2.io/#/io) plays an analogous role.

**Comment by DamianReeves** [comment]
I'd be happy to pile onto this. Are there examples of running apps that use ZIO 2? I had some issues with linking when I tried to get a simple ZIO 2 SN app running.

**Comment by sideeffffect** [comment]
Why don't you share the details about the errors you faced with us?
Perhaps in a new issue? Please mention me 🙏
I don't have any ZIO 2 working examples unfortunately.

**Comment by palanga** [comment]
> I'd be happy to pile onto this. Are there examples of running apps that use ZIO 2? I had some issues with linking when I tried to get a simple ZIO 2 SN app running.

I have a minimal example of using ZIO 2 with scala native: [https://github.com/palanga/zio-2-scala-3-native](https://github.com/palanga/zio-2-scala-3-native)

**Comment by sideeffffect** [comment]
After all the time, this can finally be put to rest :relaxed: 
@adamgfraser has fixed it for good in https://github.com/zio/zio/pull/8080

**Comment by LeeTibbert** [comment]
@adamgfraser @sideeffffect 
Congratulations and thank you for bringing this to completion. Years of effort.
I look forward to trying it out. 

Sorry I never got back to working on this, Life happened.
Part of that effort should make the underlying Scala Native more stable,
so there is/was some payoff for ZIO.

Later: @palanga  I missed thanking your efforts, sorry.

**Comment by miroslavmachura** [comment]
Happy too see this moving, thanks for your efforts!

If anyone would like to try this out - here's yet another simple example with scala-cli: https://gist.github.com/miroslavmachura/e9e3c625432eec04b33c0fad4c045a2a
